src/dask_awkward/lib/optimize.py

```diff
@@ -111,10 +111,30 @@ def optimize_columns(dsk: HighLevelGraph, keys: Sequence[Key]) -> HighLevelGraph
             all_reps.update(getattr(dsk.layers[ln].meta, "_report", ()))
     name = tokenize("output", lays)
     [_.commit(name) for _ in all_reps]
+    all_layers = tuple(dsk.layers) + (name,)
+
+    for k, lay, cols in _optimize_columns(dsk.layers, all_layers):
+        new_lay = lay.project(cols)
+        dsk2[k] = new_lay
+
+    return HighLevelGraph(dsk2, dsk.dependencies)
+
+
+def _buf_to_col(s):
+    return (
+        s[2:]
+        .replace(".content", "")
+        .replace("-offsets", "")
+        .replace("-data", "")
+        .replace("-index", "")
+        .replace("-mask", "")
+    )
```

Comment:
I don't have time *right now* to look at the full PR here, but is this assumption about buffer name mappings enforced anywhere? Right now the design is such that input sources can use a non-canonical buffer naming scheme.

For coffea I think we made this possible so that Coffea could directly pass in their own buffer names with semantic meaning.

Reply:
Do we have access to the call to `ak.from_buffers` here? That function takes a callable that constructs the buffer names from the `form_key`s and these roles.

The roles are fixed words like this, and Coffea only has the power to change the `form_key`, not the role. This is almost the complete set of such names; only `-tags` (for UnionArray) is missing.

Reply:
@jpivarski I started writing a reply here, but I think I should just aim for explicit verbosity, and hope we end up on the same page.

@martindurant this might be useful too, not sure.

When writing the current column optimisation API for dask-awkward, I had the following intentions:
- be generic enough that coffea doesn't have to write super weird code
- make few assumptions about optimisation, let the IO func do the work

Crucially, the optimisation strategies differ between Parquet, ROOT, and remapped-ROOT:
- Parquet cannot read the metadata arrays (so we have a pick-a-child strategy)
- ROOT cannot read the metadata arrays (so we have a read-the-whole-branch strategy)
- Remapped-ROOT can read individual buffers (as a single branch may correspond to an offsets buffer

To handle all of this complexity, I made column optimisation in Dask Awkward a thin contract; the strategy for optimisation remains an implementation-specific feature. We implement an optimisation strategy for "column" (pick-one-child) formats, but not for the ROOT/Remapped-ROOT cases.
The contract imposed upon IO functions wishing to implement a strategy is:

- `IOFunc.prepare_for_projection()` returns:
  - `array` - an Awkward `Array` that has buffers already named
  - `report` - a `TypeTracerReport`
  - `state` - additional state required later by `project()`
- `IOFunc.project(report, state)` returns:
  - `layer` - the projected layer


As such, Dask Awkward doesn't need to know about buffer / form keys in its optimisation machinery.

I think where possible we should keep this "black-box" approach, and have the IO-func perform the strategy for optimisation.